core/src/main/java/io/grpc/MethodDescriptor.java

```diff
@@ -262,4 +269,40 @@ public static String extractFullServiceName(String fullMethodName) {
     }
     return fullMethodName.substring(0, index);
   }
+
+  /**
+   * An internal only class for passing information directly into the transport.
+   */
+  @Internal
+  public static final class InternalAccessor {
+    private static final String PACKAGE = InternalAccessor.class.getPackage().getName();
+
+    private static final Set<String> PACKAGE_WHITELIST =
+        new HashSet<String>(Arrays.asList(PACKAGE + ".netty."));
+
+    /**
+     * Only packages on the whitelist may access this.
+     */
+    @Internal
+    public InternalAccessor() {
+      StackTraceElement[] st = new Throwable().getStackTrace();
```

Comment:
I think I'd be okay with just exposing the method with `@Internal`. If we want a stronger approach, maybe have an `Internal` class instead that provides accessors like this? That is then more obvious to the callers they are doing something they probably shouldn't. Such a class doesn't actively prohibit using it (although it could, with the same trick here), but maybe doing so is too strong?

Looking at the class name wouldn't work too well on Android, although, yes, only Netty is involved in this case.